pydantic/_internal/_std_types_schema.py

```diff
@@ -66,26 +69,42 @@ def to_enum(__input_value: Any, **_kwargs: Any) -> Enum:
         except ValueError:
             raise PydanticCustomError('enum', 'Input is not a valid enum member')
 
-    literal_schema = core_schema.literal_schema(*[m.value for m in enum_type.__members__.values()])
+    literal_schema = core_schema.literal_schema(
+        *[m.value for m in enum_type.__members__.values()],
+        ref=f'{getattr(enum_type, "__module__", None)}.{enum_type.__name__}',
+    )
+    override = literal_schema.copy()
+    json_schema_extra = JsonSchemaExtra(
+        title=enum_type.__name__, description=inspect.cleandoc(enum_type.__doc__ or 'An enumeration.')
+    )
+    override['metadata'] = build_core_metadata_for_json_schema(
+        source_class=enum_type,
+        extra=json_schema_extra,
+    )
+    metadata = build_core_metadata_for_json_schema(override_core_schema=override)
 
-    if issubclass(enum_type, IntEnum):
+    if issubclass(enum_type, int):
+        # this handles IntEnum
+        json_schema_extra.extra_updates = {'type': 'integer'}
         return core_schema.chain_schema(
-            core_schema.int_schema(), literal_schema, core_schema.function_plain_schema(to_enum)
+            core_schema.int_schema(), literal_schema, core_schema.function_plain_schema(to_enum), metadata=metadata
         )
     elif issubclass(enum_type, str):
+        # This _should_ handle 3.11's StrEnum --
+        # TODO: add test for StrEnum in 3.11, and also for enums that inherit from str/int
+        json_schema_extra.extra_updates = {'type': 'string'}
         return core_schema.chain_schema(
-            core_schema.str_schema(), literal_schema, core_schema.function_plain_schema(to_enum)
+            core_schema.str_schema(), literal_schema, core_schema.function_plain_schema(to_enum), metadata=metadata
         )
     else:
-        return core_schema.function_after_schema(
-            literal_schema,
-            to_enum,
-        )
+        return core_schema.function_after_schema(literal_schema, to_enum, metadata=metadata)
 
 
 @schema_function(Decimal)
 def decimal_schema(_schema_generator: GenerateSchema, _decimal_type: type[Decimal]) -> core_schema.FunctionSchema:
     decimal_validator = _validators.DecimalValidator()
+    metadata = build_core_metadata(update_schema=decimal_validator.__pydantic_update_schema__)
```

Comment:
Is this a good moment to mention https://github.com/pydantic/pydantic/issues/1511 ? i.e. change `float` and `int` by `str`.